src/cloudevents/core/formats/json.py

```diff
@@ -0,0 +1,74 @@
+#  Copyright 2018-Present The CloudEvents Authors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+
+import base64
+import re
+from datetime import datetime
+from json import JSONEncoder, dumps
+from typing import Any, Final, Pattern, Union
+
+from cloudevents.core.base import BaseCloudEvent
+from cloudevents.core.formats.base import Format
+
+
+class _JSONEncoderWithDatetime(JSONEncoder):
+    """
+    Custom JSON encoder to handle datetime objects in the format required by the CloudEvents spec.
+    """
+
+    def default(self, obj: Any) -> Any:
+        if isinstance(obj, datetime):
+            dt = obj.isoformat()
+            # 'Z' denotes a UTC offset of 00:00 see
+            # https://www.rfc-editor.org/rfc/rfc3339#section-2
+            if dt.endswith("+00:00"):
+                dt = dt.removesuffix("+00:00") + "Z"
+            return dt
+
+        return super().default(obj)
+
+
+class JSONFormat(Format):
+    CONTENT_TYPE: Final[str] = "application/cloudevents+json"
+    JSON_CONTENT_TYPE_PATTERN: Pattern[str] = re.compile(
+        r"^(application|text)\\/([a-zA-Z]+\\+)?json(;.*)*$"
+    )
+
+    def read(self, data: Union[str, bytes]) -> BaseCloudEvent:
+        pass
+
+    def write(self, event: BaseCloudEvent) -> bytes:
+        """
+        Write a CloudEvent to a JSON formatted byte string.
+
+        :param event: The CloudEvent to write.
+        :return: The CloudEvent as a JSON formatted byte array.
+        """
+        event_data = event.get_data()
+        event_dict: dict[str, Any] = {**event.get_attributes()}
+
+        if event_data is not None:
+            if isinstance(event_data, (bytes, bytearray)):
+                event_dict["data_base64"] = base64.b64encode(event_data).decode("utf-8")
+            else:
+                datacontenttype = event_dict.get(
+                    "datacontenttype", JSONFormat.CONTENT_TYPE
+                )
+                if re.match(JSONFormat.JSON_CONTENT_TYPE_PATTERN, datacontenttype):
+                    event_dict["data"] = dumps(event_data)
+                else:
+                    event_dict["data"] = str(event_data)
+
+        return dumps(event_dict, cls=_JSONEncoderWithDatetime).encode("utf-8")
```

Comment:
one of the things that are done not great in the v1 is that it's not possible to make e.g. JSON parsing/dumping configurable. It's often desirable to be able to e.g. change the underlying JSON library. that's smth that the new version should support out of the box.

Reply:
I'm not sure that this is a good idea. What would be the reasons? If we implement the spec, then the library should be opinionated about this. How would we guarantee that the "injected" JSON library follows the spec? 
IMO, if someone would want to have a new JSON format, they could subclass the `Format` protocol and implement `read` and `write` methods, no?

Reply:
the very first thing people would want to configure is the `cls` - the encoder. It's almost a 100% guarantee while people are often setting custom parameters to the attributes. 

I believe that we should be providing the extension points with meaningful defaults while keeping the way for people to extend the behavior when desirable. We're not doing some low-level byte-per-byte encoding.

Also depending on how this class is going to be used we may leave other extension points.

Reply:
I think this really depends on what kind of interface we want to expose here. As per the specs, the custom extensions should be serialized with the same rules as other attributes, so I would expect one to implicitly set the values of the extensions to be primitive types and not require custom JSON serialization logic here. Honestly, I am reluctant to have the encoder being a custom implementation. You can see that even right now, it's a custom one as we need to support the spec and have `time` attribute follow the ISO, and having users overwrite this, it will break the specs.

To me, if one needs to have a custom JSON encoder, then one really knows what he is doing, and having to subclass `Format` class and overwrite it is a good enough option for them. I tend to believe that we need to expose a simpler interface rather than a complex and "flexible" one, which then other things will creep in, and it's going to be a pain to maintaining the sdk. I don't think other SDKs have this capability, but they do expose interfaces that can be implemented by the end users to implement custom behaviour. 

Do you think having a callback that can be called right before the `dump` would be good enough? Something like
```python
def write(self, event: BaseCloudEvent, custom_processor: Callable) -> bytes:
    ...
    processed_dict = custom_processor(event_dict)
    return dumps(processed_dict, cls=_JSONEncoderWithDatetime).encode("utf-8")
```

Reply:
In a comment below you were referring to the usage being:

```
to_structured(event, format):
```

Format here is the format implementation instance? Depending on how we envision the end case usage for the library users we should provide extension points or overrides respectfully. I can't say whether your suggestion is good or not while I don't have a context that you have in your mind and I don't want to enforce this or that approach yet. Trying to work together to get to smth that makes sense for everyone

Reply:
> Format here is the format implementation instance?

Yes, that's right. 

The vision I have for the SDK is to separate the formats and bindings, such that each bindings implementation could work with different formats. Say I want to have a Kafka message in structured mode, I would have something like:
```python
from bindings.kafka import to_structured, from_structured
from format.json import JSONFormat
from format.avro import AvroFormat

json_format = JSONFormat()
avro_format = AvroFormat()

event = CloudEvent(...)

kafka_message_json = to_structured(event, json_format)
kafka_message_avro = to_structured(event, avro_format)
...

event_bytes: bytes = # somehow we get the message

ce_from_kafka_json = from_structured(event_bytes, json_format)
ce_from_kafka_avro = from_structured(event_bytes, avro_format)

```
By having the formats decoupled from the bindings, we allow developers write their own formats and just inject them into the bindings implementations. 

That's mostly how I see the SDK working. For sure we can go the extra mile, in a separate iteration maybe, and have  
format providers and make the `from_structured` functions smart enough that it would automatically detect the format and try to decode the message. This would a very nice improvement, and it will improve the DX by a lot.